### PR TITLE
Factorize code in NiftiLabelsMasker and NiftiMapsMasker

### PR DESCRIPTION
**Pull request by AlexandreAbraham**
First step toward code factorization in Nifti*Masker.

### CODE REVIEW

**Comment by bthirion** [comment]
You should probably address the Travis failures before we look at it in more detail ?

File: nilearn/input_data/base_masker.py [hunk 1]
```diff
@@ -270,3 +270,51 @@ def _check_fitted(self):
...
+            print("Smoothing images")
+        imgs = cache(
+            image.smooth_img, memory, func_memory_level=2,
+            memory_level=memory_level)(
+                imgs, smoothing_fwhm)
+
+    if verbose > 0:
+        print("Extracting region signals")
+    region_signals, aux = cache(extraction_func, memory, func_memory_level=2,
+                           memory_level=memory_level)(imgs)
```
Comment (banilo) [inline nilearn/input_data/base_masker.py]: PEP8

Comment (AlexandreAbraham) [inline nilearn/input_data/base_masker.py]: Fixed.

**Comment by AlexandreAbraham** [comment]
@lesteve ?

File: nilearn/input_data/nifti_maps_masker.py [hunk 2]
```diff
@@ -292,18 +246,28 @@ def transform(self, imgs, confounds=None):
                         target_shape=ref_img.shape[:3],
                         target_affine=ref_img.get_affine())
 
+        def extraction_function(imgs):
```
Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: Wow can you get away with that, without a self and without having it a staticmethod or classmethod? I thought the first argument had to be an instance of the class from the method ...

Also can you use a consistent name. You used extraction_function here and extraction_func somewhere else.

File: nilearn/input_data/base_masker.py [hunk 3]
```diff
@@ -270,3 +270,51 @@ def _check_fitted(self):
             raise ValueError('It seems that %s has not been fitted. '
                              'You must call fit() before calling transform().'
                              % self.__class__.__name__)
+
+
+# XXX This function is converging toward filter_and_mask. They will merge
+# eventually.
+def filter_and_extract(imgs, extraction_func,
```
Comment (bthirion) [inline nilearn/input_data/base_masker.py]: If it's no longer a private function, it deserves docstrings

Comment (AlexandreAbraham) [inline nilearn/input_data/base_masker.py]: This is as private as `filter_and_mask`. I added a minimal docstring though.

File: nilearn/input_data/nifti_maps_masker.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: It's a local function.

Naming fixed.

Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: > It's a local function.

Ah I missed this horizontal alignment subtlety completely in the diff.

Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: Is it good to go now? ;)

File: nilearn/input_data/nifti_maps_masker.py [hunk 4]
```diff
@@ -292,18 +246,28 @@ def transform(self, imgs, confounds=None):
...
                 # Pre-treatments
                 self.smoothing_fwhm, self.t_r, self.standardize, self.detrend,
                 self.low_pass, self.high_pass, confounds,
                 # Caching
                 self.memory, self.memory_level,
                 # kwargs
-                mask_img=self._resampled_mask_img_,
-                resample_on_maps=(self.resampling_target != 'data'),
+                target_fov=target_fov,
+                # mask_img=self._resampled_mask_img_,
```
Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: you can remove this comment, right?

File: nilearn/input_data/base_masker.py [hunk 5]
```diff
@@ -270,3 +270,62 @@ def _check_fitted(self):
...
+
+    Parameters
+    ----------
+    imgs: images
+        Images to be masked
+
+    extraction_func: function
+        Function used to extract the time series from 4D data.
+
+    kwargs: dict
```
Comment (lesteve) [inline nilearn/input_data/base_masker.py]: There is no kwargs. Maybe you mean to say something like "For all other parameters refer to NiftiMasker documentation"?

File: nilearn/input_data/nifti_maps_masker.py (hunk 2 above)
Comment (lesteve) [inline nilearn/input_data/nifti_maps_masker.py]: > Naming fixed.

Bonus points if you fix extraction_func in base_masker.py too.

File: nilearn/input_data/nifti_maps_masker.py [hunk 6]
```diff
@@ -292,18 +246,28 @@ def transform(self, imgs, confounds=None):
                         target_shape=ref_img.shape[:3],
                         target_affine=ref_img.get_affine())
 
+        def extraction_function(imgs):
+            return region.img_to_signals_maps(
+                imgs, self._resampled_maps_img_,
+                mask_img=self._resampled_mask_img_)
+
+        target_fov = None
+        if self.resampling_target != 'data':
```
Comment (banilo) [inline nilearn/input_data/nifti_maps_masker.py]: Perhaps a more explicit variable name for this one?

Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: Which one? We won't change the NiftiMapsMasker interface...

File: nilearn/input_data/nifti_maps_masker.py (hunk 2 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/nifti_maps_masker.py]: Fixed.

File: nilearn/input_data/base_masker.py (hunk 5 above)
Comment (AlexandreAbraham) [inline nilearn/input_data/base_masker.py]: Is it sphinx compatible?

Comment (AlexandreAbraham) [inline nilearn/input_data/base_masker.py]: Fixed.

File: nilearn/input_data/base_masker.py [hunk 7]
```diff
@@ -270,3 +270,61 @@ def _check_fitted(self):
...
+                       confounds, memory, memory_level,
+                       target_fov=None, verbose=0):
+    """Extract representative time series using given function.
+
+    Parameters
+    ----------
+    imgs: images
+        Images to be masked
+
+    extraction_func: function
```
Comment (lesteve) [inline nilearn/input_data/base_masker.py]: So close and yet just slightly off the mark for the bonus points ;-). Still an extraction_func to go here and I'll merge this one.

**Pull request merged**